### ISSUE TITLE
Error with set_power 0

### ISSUE THREAD

**Issue by KarlB59** [fredericvl/py-agua-iot#3]
Hi Fredericvl,

Thx for your code, it's perfect !
I've a little trouble, when i want set the power of my evacalor to 0, i've error (see below)

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/py_agua_iot/__init__.py", line 590, in set_power
    self.__request_writing(item, values)
  File "/usr/local/lib/python3.7/dist-packages/py_agua_iot/__init__.py", line 475, in __request_writing
    raise Error("Error while request device writing")
py_agua_iot.Error: Error while request device writing

During handling of the above exception, another exception occurred:
```

if i choose a value of 1 at 5, i've not problem, if the value is other, i've a good bad message

`ValueError: Value must be between 0 and 5`

What do you think ?

**Comment by gzeel** [comment]
I’m having the same issue with my pellet stove. As far as I can see only power/fan modes 1,2,3,4 and 5 are supported within the code. When my pelletstove is in “silent mode”, the fan_mode in HomeAssistant is null.
Is it a code thing or isn’t it possible to set it to fan_mode null in the API?

**Comment by fredericvl** [comment]
@KarlB59
Are you setting the power value directly via the Python module or via Home Assistant?

In the background I retrieve the minimum and maximum power values from the Agua IOT API and use that values to check if the value is in the correct range.
As the error message states between 0 and 5, a "0" should work directly via the Python module.
If not, could you please enable the debug parameter and send me all the outputs?
But you are trying to change it via Home Assistant it is another story because that one uses fixed power values from 1 to 5 (and results into seeing "null" when it is actually zero).

I've released v0.0.5 where I expose the min/max power values as an attribute in the Python module and I've also adjusted the HA example to adopt these power values and dynamically configure the possible values as fan modes depending on your stove.


@gzeel
Since I know you use HA, can you please test the new HA example version?
Fan mode "0" should work in that one if your stove supports it.

**Comment by KarlB59** [comment]
@fredericvl, thank you very much for your help.
I use directly python code (i've created a simple script for memorie all parameters (user/password/etc...) and i specify the control of my evacalor when launch my script)

I upgraded your code on v0.0.5 and enabled debug mode.
I run script with the 0 power setting and you can see this information. (i replace token/id_device and id_product by HIDE)

```
send: b'POST /deviceRequestWriting HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nlocal: false\r\nAuthorization: HIDE\r\nContent-Length: 218\r\n\r\n'
send: b'{"id_device": "HIDE", "id_product": "HIDE", "Protocol": "RWMSmaster", "BitData": [8], "Endianess": ["L"], "Items": [32976], "Masks": [65535], "Values": 0}'
reply: 'HTTP/1.1 422 Unprocessable Entity\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:35:19 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 86
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"56-Mv+wZetEMpvgrBG7HMPFb+c46uc"
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/py_agua_iot/__init__.py", line 598, in set_power
    self.__request_writing(item, values)
  File "/usr/local/lib/python3.7/dist-packages/py_agua_iot/__init__.py", line 475, in __request_writing
    raise Error("Error while request device writing")
py_agua_iot.Error: Error while request device writing

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "evacalor.py", line 59, in <module>
    device.set_power = args.valeurP
  File "/usr/local/lib/python3.7/dist-packages/py_agua_iot/__init__.py", line 600, in set_power
    raise Error("Error while trying to set power")
py_agua_iot.Error: Error while trying to set power

```

And now, you have the debug with the power = 3 (and it run corretly)


```
send: b'POST /deviceRequestWriting HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nlocal: false\r\nAuthorization: HIDE\r\nContent-Length: 218\r\n\r\n'
send: b'{"id_device": "HIDE", "id_product": "HIDE", "Protocol": "RWMSmaster", "BitData": [8], "Endianess": ["L"], "Items": [32976], "Masks": [65535], "Values": 3}'
reply: 'HTTP/1.1 401 Unauthorized\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:46:39 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 54
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"36-Fwe8VDbE+GiEZRCgh42kIkXKsFw"
send: b'POST /refreshToken HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nContent-Length: 386\r\n\r\n'
send: b'{"refresh_token": "HIDE"}'
reply: 'HTTP/1.1 201 Created\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:46:39 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 432
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"1b0-5xlDQLkt2Syat2nhpGNDkKc/gwg"
send: b'POST /deviceRequestWriting HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nlocal: false\r\nAuthorization: HIDE\r\nContent-Length: 218\r\n\r\n'
send: b'{"id_device": "HIDE", "id_product": "HIDE", "Protocol": "RWMSmaster", "BitData": [8], "Endianess": ["L"], "Items": [32976], "Masks": [65535], "Values": 3}'
reply: 'HTTP/1.1 200 OK\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:46:39 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 102
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"66-/afThLFXDXHQ57tkBvtHe058PUM"
send: b'GET /deviceJobStatus/a502ba4d-bdd7-4a5e-946a-c3f8be1088a1 HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nlocal: false\r\nAuthorization: HIDE\r\nContent-Length: 2\r\n\r\n'
send: b'{}'
reply: 'HTTP/1.1 200 OK\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:46:39 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 118
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"76-+PremWLcB6OiOFBFLZa+qQez4sQ"
send: b'GET /deviceJobStatus/a502ba4d-bdd7-4a5e-946a-c3f8be1088a1 HTTP/1.1\r\nHost: jollymec.agua-iot.com\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: application/json, text/javascript, */*; q=0.01\r\nConnection: keep-alive\r\nContent-Type: application/json\r\nOrigin: file://\r\nid_brand: 1\r\ncustomer_code: 732584\r\nlocal: false\r\nAuthorization: HIDE\r\nContent-Length: 2\r\n\r\n'
send: b'{}'
reply: 'HTTP/1.1 200 OK\r\n'
header: Server: nginx/1.10.3 (Ubuntu)
header: Date: Sat, 06 Feb 2021 14:46:40 GMT
header: Content-Type: application/json; charset=utf-8
header: Content-Length: 352
header: Connection: keep-alive
header: X-Powered-By: Express
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, customer_code, id_brand, local, Access-Control-Allow-Origin, filepath, expect, host, content-length, enctype, Connection, Origin, Accept-Encoding, Accept-Language, User-Agent, Content-Disposition, Upgrade
header: Access-Control-Allow-Methods: GET,PUT,POST,DELETE
header: Access-Control-Allow-Credentials: true
header: ETag: W/"160-78gSeCk5gwe8q6l8h7Yse8QZAcY"
```

i've test your new command, min and max power et you can see than the power 0 it's corretly detected

```
min_P:0
max_P:5
```

**Comment by fredericvl** [comment]
@KarlB59
Thanks for the debug logs!

Looks like the Agua IOT API is returning `422 Unprocessable Entity` when sending the value "0".
I would love to help you further but I only have a friend that has a stove but his one only accepts power values from 1 to 5, so I can't really test setting the power value to 0.

Are you in the ability to sniff the packets from for example the mobile app when you set the power to 0 ?
That would help to see what exactly is being sent and then I could have a look to implement it in the module.

**Comment by gzeel** [comment]
> @gzeel
> Since I know you use HA, can you please test the new HA example version?
> Fan mode "0" should work in that one if your stove supports it.

Tested it, but I think I ran into te same bug as @KarlB59. 

Message from HA is :
`Failed to set fan mode, error: Error while trying to set power `

**Comment by fredericvl** [comment]
@KarlB59 , @gzeel
I've released v0.0.6 with some extra debug logging and also a possible fix for setting the power value to 0.
Since I cannot test it myself due to lack of a stove which supports zero power, can one of you guys please update to this version and test if setting the power now works with a value of zero?

**Comment by KarlB59** [comment]
@fredericvl 
Woah ! very good job !
it's ok for me, i've test directly via the python code and the 0 value are not error.
i ve verify on application and the "silent mode" or "0" is corretly run.
for me, this thread is terminated :) very thank again.
